man/man8/madvsnoop.8
.TH madvsnoop 8  "2022-12-05" "USER COMMANDS"
.SH NAME
madvsnoop \- Trace application memory advise issued by the madvise() syscall to the kernel. Uses Linux eBPF/bcc.
.SH SYNOPSIS
.B madvsnoop [\-h] [\-x] [-p PID] [-T PID]
.SH DESCRIPTION
madvsnoop traces the madvise() syscall, to show signals sent via this method. This
may be useful when trying to improve applications/system performance.

This works by tracing the kernel sys_madvise() function using trace point, and
will need updating to match any changes to this function.

This makes use of Linux 4.7+ (BPF_PROG_TYPE_TRACEPOINT support).

and will not work for kernels older than 4.7, mechanisms to make it work for older kernel are however in active development.

Since this uses BPF, only the root user can use this tool.
.SH REQUIREMENTS
CONFIG_BPF and bcc.
.SH OPTIONS
.TP
\-h
Print usage message.
.TP
\-x
Only print failed madvise() syscalls.
.TP
\-p PID
Trace this process ID only which is the sender of signal (filtered in-kernel).
.TP
\-T PID
Trace this target process ID only which is the receiver of signal (filtered in-kernel).
.TP
\-s SIGNAL
Trace this signal only (filtered in-kernel).
.SH EXAMPLES
.TP
Trace all madise() syscalls:
#
.B madvsnoop
.TP
Trace only madvise() syscalls that failed:
#
.B madvsnoop \-x
.TP
Trace PID 181 only:
#
.B madvsnoop \-p 181
.TP
Trace target PID 189 only:
#
.B madvsnoop \-T 189
.TP
Trace signal 9 only:
#
.B madvsnoop \-s 9
.SH FIELDS
.TP
TIME
Time of the madvise() call.
.TP
PID
Source process ID
.TP
COMM
Source process name
.TP
START
Where the memory start.
.TP
LENGTH
Memory size
.TP 
BEHAVIOUR
The type of advise
.TP
RES
Result. 0 == success, a negative value (of the error code) for failure.
.SH OVERHEAD
This traces the kernel madvise function and prints output for each event. As the
rate of this is generally expected to be low (< 100/s), the overhead is also
expected to be negligible. If you have an application that is calling a very
high rate of madvise()s for some reason, then test and understand overhead before
use.
.SH SOURCE
This is from bcc.
.IP
https://github.com/iovisor/bcc
.PP
Also look in the bcc distribution for a companion _examples.txt file containing
example usage, output, and commentary for this tool.
.SH OS
Linux
.SH STABILITY
Unstable - in development.
.SH AUTHOR
Jules Irenge
.SH SEE ALSO
shmsnoop(8), memleak(8)


tools/madvsnoop.py
# !/usr/bin/python
#
# madvsnoop   Trace mdavise syscall displaying values from every programs.
#             For Linux, uses BCC, eBPF. Embedded C.
#
# USAGE: madvsnoop [-h] [-p PID]
# REQUIRES: Linux 4.7+ (BPF_PROG_TYPE_TRACEPOINT support).
#
#
# Copyright 2022 ELISA, Linux Foundation Mentorship program
# Licensed under the Apache License, Version 2.0 (the "License")

# 15-Nov-2022   Jules Irenge   Created this, from borrowed template

from __future__ import print_function
from bcc import BPF
from bcc.utils import printb
import argparse
import signal
from time import strftime
import os, subprocess

# arguments
examples = """examples:
    ./madvsnoop           # trace all madvise() signals
    ./madvsnoop -p 4086    # only trace PID 4086
"""
parser = argparse.ArgumentParser(
    description="Trace signals issued by the madvise() syscall",
    formatter_class=argparse.RawDescriptionHelpFormatter,
    epilog=examples)
parser.add_argument("-x", "--failed", action="store_true",
    help="only show failed madvise syscalls")
parser.add_argument("-p", "--pid",
    help="trace this PID only")
parser.add_argument("--ebpf", action="store_true",
    help=argparse.SUPPRESS)
args = parser.parse_args()
debug = 0

# define BPF program
bpf_text = """

#include <uapi/linux/ptrace.h>
#include <linux/sched.h>

struct val_t {
        u64 pid;
        int start;
        int length;
        int behaviour;
        char comm[TASK_COMM_LEN];
        };
struct data_t {
        u64 pid;
        int start;
        int length;
        int behaviour;
        char comm[TASK_COMM_LEN];
        int ret;
        };
BPF_HASH(infotmp, u32, struct val_t);
BPF_PERF_OUTPUT(events);

TRACEPOINT_PROBE(syscalls, sys_enter_madvise) {
u64 pid_tgid = bpf_get_current_pid_tgid();
u32 pid = pid_tgid >> 32;
u32 tid = (u32) pid_tgid;

PID_FILTER

struct val_t val = {.pid = pid};

if(bpf_get_current_comm(&val.comm, sizeof(val.comm)) == 0) {
        val.start = args->start;
        val.length = args->len_in;
        val.behaviour = args->behavior;

        infotmp.update(&tid, &val);
        }
        return 0;
        }
TRACEPOINT_PROBE(syscalls, sys_exit_madvise) {
struct data_t data = {};
struct val_t *valp;
u64 pid_tgid = bpf_get_current_pid_tgid();
u32 pid = pid_tgid >> 32;
u32 tid = (u32) pid_tgid;

valp = infotmp.lookup(&tid);
if(valp == 0) {
//missed entry
return 0;
}
bpf_probe_read_kernel(&data.comm, sizeof(data.comm), valp->comm);
data.pid = pid;
data.start = valp->start;
data.length = valp->length;
data.behaviour = valp->behaviour;
data.ret =  args->ret;
void *ctx = args;

events.perf_submit(ctx, &data, sizeof(data));

return 0;

        }
"""
if os.geteuid() != 0:
    exit("You need to have root privileges to run this script.\nPlease try again, this time using 'sudo'. Exiting.")

if args.pid:
    bpf_text = bpf_text.replace('PID_FILTER',
        'if (pid != %s) { return 0; }' % args.pid)
else:
    bpf_text = bpf_text.replace('PID_FILTER', '')
if debug or args.ebpf:
    print(bpf_text)
    if args.ebpf:
        exit()


# initialize BPF
b = BPF(text=bpf_text)

print("%-9s %-6s %-16s %-16s %-9s %-9s %3s" % (
    "TIME", "PID", "COMM", "START", "LENGTH", "BEHAVIOUR", "RESULT"))

# process event
def print_event(cpu, data, size):
    event = b["events"].event(data)

    if (args.failed and (event.ret >= 0)):
        return

    printb(b"%-9s %-6d %-16s %-16d %-9d %-9d %3d" %
           (strftime("%H:%M:%S").encode('ascii'),
            event.pid, event.comm, event.start,
            event.length, event.behaviour, event.ret))

# loop with callback to print_event
b["events"].open_perf_buffer(print_event)
while 1:
    try:
        b.perf_buffer_poll()
    except KeyboardInterrupt:
        exit()
